### PR TITLE
[CALCITE-3914] Improve SubsitutionVisitor to consider RexCall of type PLUS and TIMES for canonicalization (Vineet Garg)

### PR DESCRIPTION
**Pull request by vineetgarg02**
…

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 1]
```diff
@@ -324,6 +324,8 @@ public static RexNode splitFilter(final RexSimplify simplify,
    */
   private static RexNode canonizeNode(RexBuilder rexBuilder, RexNode condition) {
     switch (condition.getKind()) {
+    case PLUS:
```
Comment (jcamachor) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:327]: The code in this block removes duplicate operands (via the `TreeMap`). You will need slightly different code for `PLUS` or `TIMES` (just the sort piece).

Once this is fixed, could you add a test with duplicate operand, e.g,., `x + x` ?

Comment (vineetgarg02) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java]: @jcamachor Thanks for catching this. I have updated the pull request.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 2]
```diff
@@ -345,13 +345,29 @@ private static RexNode canonizeNode(RexBuilder rexBuilder, RexNode condition) {
     case LESS_THAN_OR_EQUAL:
     case GREATER_THAN_OR_EQUAL: {
       RexCall call = (RexCall) condition;
-      final RexNode left = call.getOperands().get(0);
-      final RexNode right = call.getOperands().get(1);
+      RexNode left = call.getOperands().get(0);
```
Comment (jcamachor) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:348]: nit. L348 and L349 could be merged into a single one. Same below.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 3]
```diff
@@ -345,13 +345,29 @@ private static RexNode canonizeNode(RexBuilder rexBuilder, RexNode condition) {
...
+      RexCall call = (RexCall) condition;
+      RexNode left = call.getOperands().get(0);
+      RexNode right = call.getOperands().get(1);
+      left = canonizeNode(rexBuilder, left);
+      right = canonizeNode(rexBuilder, right);
+
+      if (left.toString().compareTo(right.toString()) <= 0) {
+        return rexBuilder.makeCall(call.getOperator(), left, right);
+      } else {
+        return rexBuilder.makeCall(call.getOperator(), right, left);
```
Comment (jcamachor) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:369]: We should introduce an additional check before returning the new swapped expression to verify that the return type of the new call is similar to the return type of the original one (we can do that via `SqlTypeUtil.equalSansNullability`); otherwise, we should not swap. For instance, in Calcite the type inference for + or * will be dependent on the order of the operands. But other systems, e.g., Hive, may have their own implementation and apply different rules for type inference.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 4]
```diff
@@ -345,13 +345,29 @@ private static RexNode canonizeNode(RexBuilder rexBuilder, RexNode condition) {
...
       RexCall call = (RexCall) condition;
-      final RexNode left = call.getOperands().get(0);
-      final RexNode right = call.getOperands().get(1);
+      RexNode left = call.getOperands().get(0);
+      left = canonizeNode(rexBuilder, left);
+      RexNode right = call.getOperands().get(1);
+      right = canonizeNode(rexBuilder, right);
+      call = (RexCall) rexBuilder.makeCall(call.getOperator(), left, right);
       if (left.toString().compareTo(right.toString()) <= 0) {
         return call;
```
Comment (jcamachor) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:354]: You can only return the original call if `left==call.getOperands().get(0) && right==call.getOperands().get(1)`. Below (for + and *), you could check if that is the case too, and return the original call otherwise; it is worth skipping new expression creation in these cases).
